sound/soc/sof/pm.c

```diff
@@ -411,6 +427,39 @@ int snd_sof_runtime_resume(struct device *dev)
 }
 EXPORT_SYMBOL(snd_sof_runtime_resume);
 
+int snd_sof_set_d0_state(struct snd_sof_dev *sdev,
+			 enum sof_d0_substate d0_substate)
+{
+	u32 flags;
+	int ret;
+
+	dev_dbg(sdev->dev, "setting ADSP D0 substate:%d\n", d0_substate);
+
+	/* Todo: destroy/reload unused pipelines/widgets in FW */
+
+	/* do platform specific set_state */
+	ret = snd_sof_dsp_set_power_state(sdev, d0_substate, &flags);
+	if (ret < 0) {
+		dev_err(sdev->dev,
+			"error: d0i3_exit ipc error %d\n", ret);
+		return ret;
+	}
+
+	/* sending pm_gate IPC */
```

Comment:
if there is no platform-specific dsp_set_power_state(), then flags is not initialized, and the subsequent pm_gate_ipc makes no sense.

You should do the pm_gate_ipc within the HDaudio layers.

Reply:
yes, thanks for spotting that, let me move this part to platform specific.